lib/auth/keystore/gcp_kms.go

```diff
@@ -0,0 +1,454 @@
+	clock           clockwork.Clock
+}
+
+// newGCPKMSKeyStore returns a new keystore configured to use a GCP KMS keyring
+// to manage all key material.
+func newGCPKMSKeyStore(cfg *GCPKMSConfig) (*gcpKMSKeyStore, error) {
+	var kmsClient kmsClientI
+	if cfg.kmsClientOverride != nil {
+		kmsClient = cfg.kmsClientOverride
+	} else {
+		realClient, err := kms.NewKeyManagementClient(context.Background())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		kmsClient = &kmsClientWrapper{realClient}
+	}
+
+	clock := clockwork.NewRealClock()
+	if cfg.clockOverride != nil {
+		clock = cfg.clockOverride
+	}
+
+	return &gcpKMSKeyStore{
+		hostUUID:        cfg.HostUUID,
+		keyRing:         cfg.KeyRing,
+		protectionLevel: gcpKMSProtectionLevels[cfg.ProtectionLevel],
+		kmsClient:       kmsClient,
+		log:             logrus.WithFields(logrus.Fields{trace.Component: "GCPKMSKeyStore"}),
+		clock:           clock,
+	}, nil
+}
+
+func newKeyID() (string, error) {
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return id.String(), nil
+}
+
+// generateRSA creates a new RSA private key and returns its identifier and a
+// crypto.Signer. The returned identifier for gcpKMSKeyStore encoded the full
+// GCP KMS key version name, and can be passed to getSigner later to get the same
+// crypto.Signer.
+func (g *gcpKMSKeyStore) generateRSA(opts ...RSAKeyOption) ([]byte, crypto.Signer, error) {
+	ctx := context.TODO()
+
+	options := &RSAKeyOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	var alg kmspb.CryptoKeyVersion_CryptoKeyVersionAlgorithm
+	switch options.DigestAlgorithm {
+	case crypto.SHA256, 0:
+		alg = kmspb.CryptoKeyVersion_RSA_SIGN_PKCS1_2048_SHA256
+	case crypto.SHA512:
+		alg = kmspb.CryptoKeyVersion_RSA_SIGN_PKCS1_4096_SHA512
+	default:
+		return nil, nil, trace.BadParameter("unsupported digest algorithm: %v", options.DigestAlgorithm)
+	}
+
+	keyUUID, err := newKeyID()
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	req := &kmspb.CreateCryptoKeyRequest{
+		Parent:      g.keyRing,
+		CryptoKeyId: keyUUID,
+		CryptoKey: &kmspb.CryptoKey{
+			Purpose: kmspb.CryptoKey_ASYMMETRIC_SIGN,
+			Labels: map[string]string{
+				hostLabel: g.hostUUID,
+			},
+			VersionTemplate: &kmspb.CryptoKeyVersionTemplate{
+				ProtectionLevel: g.protectionLevel,
+				Algorithm:       alg,
+			},
+		},
+	}
+	resp, err := doGCPRequest(ctx, g.clock, g.kmsClient.CreateCryptoKey, req)
+	if err != nil {
+		return nil, nil, trace.Wrap(err, "error while attempting to generate new GCP KMS key")
+	}
+
+	keyID := gcpKMSKeyID{
+		keyVersionName: resp.Name + keyVersionSuffix,
+	}
+
+	signer, err := g.newKmsSigner(ctx, keyID)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	return keyID.marshal(), signer, nil
+}
+
+// getSigner returns a crypto.Signer for the given pem-encoded private key.
+func (g *gcpKMSKeyStore) getSigner(rawKey []byte) (crypto.Signer, error) {
+	ctx := context.TODO()
```

Comment:
Same question as with `generateRSA`, why not just take a `context` and let the caller pass in a `TODO` if nothing else is available?